### ISSUE TITLE
ActivityPub: POSTing to inbox causes 400 Bad Request

### ISSUE THREAD

**Issue by andrigamerita** [NodeBB/NodeBB#12708]
### NodeBB version

v4.0.0-alpha

### NodeBB git hash

fc5a829c0cf4778840fbffdccbcd69ebdf2f1d28

### NodeJS version

v18.19.0

### Installed NodeBB plugins

	* nodebb-plugin-2factor@7.5.3 (installed, enabled)
	* nodebb-plugin-composer-default@10.2.36 (installed, enabled)
	* nodebb-plugin-dbsearch@6.2.5 (installed, enabled)
	* nodebb-plugin-emoji@5.1.15 (installed, enabled)
	* nodebb-plugin-emoji-android@4.0.0 (installed, enabled)
	* nodebb-plugin-markdown@13.0.0-pre.9 (installed, enabled)
	* nodebb-plugin-mentions@4.6.7 (installed, enabled)
	* nodebb-plugin-ns-custom-fields@6.0.0 (installed, disabled)
	* nodebb-plugin-ns-embed@6.0.0 (installed, enabled)
	* nodebb-plugin-ntfy@1.7.4 (installed, disabled)
	* nodebb-plugin-spam-be-gone@2.2.2 (installed, disabled)
	* nodebb-rewards-essentials@1.0.0 (installed, enabled)
	* nodebb-theme-harmony@2.0.0-pre.32 (installed, disabled)
	* nodebb-theme-lavender@7.1.8 (installed, disabled)
	* nodebb-theme-peace@2.2.6 (installed, disabled)
	* nodebb-theme-persona@13.3.25 (installed, enabled)
	* nodebb-widget-essentials@7.0.17 (installed, enabled)

### Database type

Redis

### Database version

Redis server v=7.0.15 sha=00000000:0 malloc=jemalloc-5.3.0 bits=64 build=8fef3e995a542118

### Exact steps to cause this issue

0. Have a NodeBB instance in this version with ActivityPub federation enabled, and ensure permissions are set to their default, meaning users of the `fediverse` group appear has having the expected, normal permissions.

Posting issues:
1. Copy the link to a NodeBB post and open it in another Fediverse platform, like Mastodon or Lemmy.
2. Send something in reply to the post in question; alternatively, specifically on Lemmy, navigate to the community page the opened post appears as belonging to (which simply corresponds to the category on NodeBB), and create a brand new thread.
3. Notice how the new post/thread created from the outside never appears on NodeBB.

The same issue is present for following users:
1. Search a NodeBB user in another Fediverse platform which supports follows, like Mastodon.
2. Try to follow the user in question.
3. Notice how no notification is sent to the followed user on NodeBB, how their follower count doesn't rise up, and how the Fediverse platform can't effectively complete the follow, which will show as pending (for Mastodon specifically, the follow won't appear as pending until refreshing the page).

Finally, this applies to both cases:

4. Notice how logs for the server hosting NodeBB (assuming a reverse proxy is being used with a sufficient logging level) show a 400 error code for the POST request associated with the action that was just carried out.

### What you expected

I expected follows to go through from a remote Mastodon instance to the NodeBB instance, as well as replies to threads (from Mastodon and Lemmy) and new threads created in a category (from Lemmy) to go trough and be displayed on NodeBB, as those are outlined as supported on the announcement thread for v4.0.0 (https://community.nodebb.org/topic/18123/nodebb-v4.0.0-alpha-aka-activitypub-alpha-release).

### What happened instead

As I can see from reading lines concerning requests from ActivityPub servers in my nginx access logs, and using https://activitypub.academy for additional testing, otherwise well-formed requests to inbox endpoints fail, always returning HTTP code 400 with body `Bad Request`, effectively meaning communication only works from NodeBB to the Fediverse, not the other way around. I can see this happening both with the general `/inbox` endpoint and the `/uid/<user id>/inbox` endpoints.

### Anything else?

Just so you know, my NodeBB is a completely clean installation done just this morning, because I initially tried to migrate some data from an old forum, but issues arose, so I gave up and decided to start fresh to avoid encountering any strange issues.

Having tested this problem with Activitypub.academy, which can log some requests, including replies to posts, I can show the following example, in which you can see how posting a reply to a post fails in the way I described.
Please note that all paths on my NodeBB instance are prefixed with `/new/`; that's intentional since, I still have an old forum that I will not be able to move until I'm certain that I can migrate to NodeBB, but I've configured both the instance and my reverse proxy correctly, and I've verified that the issue I'm reporting is present even when hosting NodeBB on the root of the same domain.

From https://activitypub.academy/users/badalia_aodrost
Sent to https://bbs.spacc.eu.org/new/inbox
Create
Note
[@octt](https://bbs.spacc.eu.org/new/user/octt) Test, prova...
20:37:08
```json
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    {
      "ostatus": "http://ostatus.org#",
      "atomUri": "ostatus:atomUri",
      "inReplyToAtomUri": "ostatus:inReplyToAtomUri",
      "conversation": "ostatus:conversation",
      "sensitive": "as:sensitive",
      "toot": "http://joinmastodon.org/ns#",
      "votersCount": "toot:votersCount"
    }
  ],
  "id": "https://activitypub.academy/users/badalia_aodrost/statuses/112837252110080163/activity",
  "type": "Create",
  "actor": "https://activitypub.academy/users/badalia_aodrost",
  "published": "2024-07-23T18:37:08Z",
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "cc": [
    "https://activitypub.academy/users/badalia_aodrost/followers",
    "https://bbs.spacc.eu.org/new/uid/2"
  ],
  "object": {
    "id": "https://activitypub.academy/users/badalia_aodrost/statuses/112837252110080163",
    "type": "Note",
    "summary": null,
    "inReplyTo": "https://bbs.spacc.eu.org/new/post/7",
    "published": "2024-07-23T18:37:08Z",
    "url": "https://activitypub.academy/@badalia_aodrost/112837252110080163",
    "attributedTo": "https://activitypub.academy/users/badalia_aodrost",
    "to": [
      "https://www.w3.org/ns/activitystreams#Public"
    ],
    "cc": [
      "https://activitypub.academy/users/badalia_aodrost/followers",
      "https://bbs.spacc.eu.org/new/uid/2"
    ],
    "sensitive": false,
    "atomUri": "https://activitypub.academy/users/badalia_aodrost/statuses/112837252110080163",
    "inReplyToAtomUri": "https://bbs.spacc.eu.org/new/post/7",
    "conversation": "tag:activitypub.academy,2024-07-23:objectId=183422:objectType=Conversation",
    "content": "<p><span class=\"h-card\"><a href=\"https://bbs.spacc.eu.org/new/user/octt\" class=\"u-url mention\">@<span>octt</span></a></span> Test, prova...</p>",
    "contentMap": {
      "en": "<p><span class=\"h-card\"><a href=\"https://bbs.spacc.eu.org/new/user/octt\" class=\"u-url mention\">@<span>octt</span></a></span> Test, prova...</p>"
    },
    "attachment": [],
    "tag": [
      {
        "type": "Mention",
        "href": "https://bbs.spacc.eu.org/new/uid/2",
        "name": "@octt@bbs.spacc.eu.org"
      }
    ],
    "replies": {
      "id": "https://activitypub.academy/users/badalia_aodrost/statuses/112837252110080163/replies",
      "type": "Collection",
      "first": {
        "type": "CollectionPage",
        "next": "https://activitypub.academy/users/badalia_aodrost/statuses/112837252110080163/replies?only_other_accounts=true&page=true",
        "partOf": "https://activitypub.academy/users/badalia_aodrost/statuses/112837252110080163/replies",
        "items": []
      }
    }
  },
  "signature": {
    "type": "RsaSignature2017",
    "creator": "https://activitypub.academy/users/badalia_aodrost#main-key",
    "created": "2024-07-23T18:37:08Z",
    "signatureValue": "fHGL1Wgy3h4ZY8TEGzWfphEN7hOoqwm83xeq/lI3IeSV7/Zy27H2Dtf6Hrt3RW0KTwpuUEw/qRsQ+ro5VIPqMI0QqHFaGQ80+GSKu3QGCCFdi+SdMMYNF7WyJpsB/EYo6+Rn8TXqIGF7D6k32rgmXVQOdZF0hhwPAxgdF5d8DfXDXo6u4Gfp05dULgn4jri6EVgPpUlpFUWM4WUOWNcji8H9w7FjLEdPpTZBRSYLyUJ5Zpa4Z6M8l0WZ8/bZIBcZgXsxFpo84WN+sSzcnF40HFFv4ieO1tNtxNJtnOLMaDwPUxpWtd2P+9Yp8MEafZeYZz7r9UdgWcAs5jhyNLzYvA=="
  }
}
```
Sending failed (https://bbs.spacc.eu.org/new/inbox responded with status 400 Bad Request)

**Comment by julianlam** [comment]
Thank you for the detailed report. 

It is possible that the default privileges as set by NodeBB on first installation are not appropriate. 

Can you please attach a screenshot of the privileges table for "global" and the category in question please?

**Comment by andrigamerita** [comment]
@julianlam Sure,

# Global
![image](https://github.com/user-attachments/assets/a387876d-069f-4988-8e96-3610b9656fbc)
![image](https://github.com/user-attachments/assets/b55c502d-b367-4c2f-bafb-7f6d671b0052)

# General Discussion
![image](https://github.com/user-attachments/assets/e82cac70-13ec-4587-b830-3d4e0017170b)
![image](https://github.com/user-attachments/assets/fbd9e960-601d-4b09-b019-ac878c2de642)

**Comment by julianlam** [comment]
Unable to see anything obviously wrong in the code or those privileges, I set up a clean install in an attempt to reproduce:

* Using Redis as data store
* Queried admin user using ActivityPub Academy (hereafter, AA) and issued a follow
* Confirmed that a notification was received and that the follower is present in the local user's follow list
* Looked up pid 1 via AA and sent a reply w/ no direct mention
* Received a notification for the reply and noted that reply was added to topic

![image](https://github.com/user-attachments/assets/a8b5f682-c2f9-4954-b714-091dadd99e60)

Sorry, at this point I cannot reproduce the issue.

**Comment by julianlam** [comment]
![image](https://github.com/user-attachments/assets/7098c3e8-a30d-4e90-b1f8-eb4c415839db)

**Comment by julianlam** [comment]
I am indeed noting that the request from AA to your instance fails:

![image](https://github.com/user-attachments/assets/7dbd4667-1f6d-42c2-bd08-8444dd8f4a81)

At this point I am not certain whether it is related to the subfolder, even though you noted in your initial issue that you tried without the `/new` suffix.

**Comment by andrigamerita** [comment]
If this is how things are then, would it be possible for the cause to be any combination of the following?
* NodeBB's ActivityPub code doesn't play nice with being on a subpath instead of the root (even if everything else does, which is strange but I wouldn't exclude it)
* Changing the configuration from the subpath to the root, then rebuilding and restarting, didn't solve the issue just because of aggressive caching on the external instances' side

Since we know that it does work for you on the root of your domain, I guess that we should do some more testing on both of these hypotheses. Any of us could try running a clean instance on a non-root path on a new domain and, if that causes the issue, then it would be a bug. If that, however, was to work, I guess the only way out of this situation will be moving my current instance to a new domain, and dealing with the uncomfortable implications... :sweat_smile: I will try testing tomorrow and report my findings.

**Comment by julianlam** [comment]
My guess is the second, that perhaps it worked fine and then you added the folder path and things didn't quite work so nicely?

However I would not be surprised if the ActivityPub code does not work on subfolder path. I did not test against it, so this is probably a good thing to investigate.

**Comment by andrigamerita** [comment]
Today I did all the experiments I could, of which only the last one was positive, and... TLDR: apparently it's an issue with the reverse proxy? but not with my configuration specifically:
* First, I used a greatly simplified nginx configuration for my real instance (removing all the rules related to the old forum still being hosted on the same domain), trying to exclude the possibility of any misconfiguration on my end, but still keeping the `/new/` prefix... no luck
* Initialized new instances on temporary domains, first with the URL prefix and then without... strangely enough, no dice
* Again, initialized a new instance on a temporary domain, on the root, but this time without any kind of HTTP proxy and... apparently this way everything worked.

Both with the nginx site configuration suggested in the docs (https://docs.nodebb.org/installing/os/debian/#configuring-nginx), and with my extremely simplified version (see below), something goes wrong. As for the server configuration itself, I'm using Debian's default `/etc/nginx/nginx.conf` on the desktop PC where I test, and the same but slightly modified on my server (to accomodate other software), and I have no idea if there's something I'm missing here, since everything else works correctly, both inside NodeBB and with other software (even other Fediverse software, when I used to host it).

```nginx
server {
  listen 80;
  # using localhost.run for testing I don't have to worry about SSL, but obviously configuration on my real server does include options for that
  location / {
    proxy_pass http://localhost:4567;
    #proxy_http_version 1.1;
    #proxy_set_header X-Real-IP $remote_addr;
    #proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    #proxy_set_header X-Forwarded-Proto $scheme;
    #proxy_set_header Upgrade $http_upgrade;
    #proxy_set_header Connection "upgrade";
  }   
}
```

At this point I'm curious on how you're running community.nodebb.org; I assume you over there are using nginx, how come you're not experiencing any issues, if Federation is enabled there?

**Comment by julianlam** [comment]
Ah, by the way, I assume you're using an SSL certificate? That's essentially a requirement for ActivityPub. I just noted that your nginx proxy didn't contain `listen 443 ssl;`

**Comment by andrigamerita** [comment]
> Ah, by the way, I assume you're using an SSL certificate?

Yes, please see my comment in the configuration file. I have all the SSL requirements on my server where I use my actual domain, but on my desktop I'm using a reverse proxy service which connects to my local HTTP port and automatically handles sending data over HTTPS. Otherwise, federation wouldn't be working at all, instead of one-way (which it does with any configuration).

---

Btw, in any case I would like to point out that there is indeed an intrinsic issue with running federation on a subpath of the domain, but it's not directly related to the problem here. In short, webfinger queries, necessary for other servers to query users by their username and get instance information, and so necessary for ActivityPub to work, fail when NodeBB is running on a subpath and no reverse proxy rule is used to specifically handle them. A bad redirect is done from `/.well-known/webfinger?resource=...` to `/subpath/.well-known/webfinger`, stripping the query parameters and as such redirecting the remote server towards making a bad request.

**Comment by julianlam** [comment]
FWIW I use nginx for the NodeBB Community, and Caddy for a couple other sites.

Let's Encrypt for all of them.

**Comment by andrigamerita** [comment]
Okay, now I'm sure there's something wrong in the interaction between nginx and NodeBB, but I'm still oblivious to why. I've verified that federation is indeed working both ways on my production NodeBB instance, still hosted on `https://bbs.spacc.eu.org`, by manually generating and then sending requests to inbox bypassing nginx. So this is not strictly a problem with NodeBB. I've done the following:

* Enabled `debug` logging for nginx
* Sent a follow request to my NodeBB profile from AA
* Copied the JSON body of the request from AA, and the HTTP headers from the nginx log
* Put everything together in a text file and sent it to NodeBB with `nc localhost 4567 < file`
* Noticed how the server responded with code 202 and body `{"status":{"code":"accepted","message":"Accepted"},"response":{}}`, and indeed on my forum's site a follow notification arrived
* Same result by assembling another file, this time with the data for an unfollow request

This is the file I constructed for the follow request (the only things that differ in the unfollow request apart from the body are the Date, Digest and Signature headers, so it's redundant to show that one too):
```
POST /new/uid/2/inbox HTTP/1.1
X-Real-IP: 10.0.3.1
X-Forwarded-For: 10.0.3.1
X-Forwarded-Proto: https
Connection: upgrade
Host: bbs.spacc.eu.org
Content-Length: 249
User-Agent: http.rb/5.1.1 (Mastodon/4.1.18; +https://activitypub.academy/)
Date: Fri, 26 Jul 2024 15:37:54 GMT
Accept-Encoding: gzip
Digest: SHA-256=obcSNZniCbWRXy/E3Ipb7gZ2sGOniQ9oDApR/rbT5/8=
Content-Type: application/activity+json
Signature: keyId="https://activitypub.academy/users/eubocus_garldeshun#main-key",algorithm="rsa-sha256",headers="(request-target) host date digest content-type",signature="KacaP3JmgGKwKFSrwnoFJTQw7QAEL0mRS2M+PT9NWy/CnW2KsTFanbvCd45OlwWsge6RzapD2/sq3DTdJbY88ww9MWPqgbOeWMnOj1lku9f3agRgk4goCWlTlglpwzJeC9hyawRhqK0mDZ7pZB/DhvMR3KraiVqLQ6NgKv6ZIcSHdu+L8sicMm4bKTLGSbgXS0hK4emBBDKvsZALmFajPuyQU+FLqbtKj/IsXNXogrhx96sNgnbzw0tHE49DmzYPOQFxOV0jHqbw6mbTwP5v0aoqBZr8+fZFOfuI1rtYIW8Y7cf4cwfG0Q5EbNRl5ZDXG7CqP6sgUMpb6/4iDmpDCA=="

{"@context":"https://www.w3.org/ns/activitystreams","id":"https://activitypub.academy/dd767dd9-d9c3-4240-a5fd-a9cfb1550331","type":"Follow","actor":"https://activitypub.academy/users/eubocus_garldeshun","object":"https://bbs.spacc.eu.org/new/uid/2"}

```

It's interesting to note that I do get the 400 Bad Request error if sending this exact data, but to nginx instead (`nc bbs.spacc.eu.org 80`; yes, it works on plain HTTP since I have nginx configured to support both HTTPS and HTTP connections without forcing either one), as we've continuously observed; however, if I first send the data directly to NodeBB as noted above, and then send the exact same data to the server via nginx, I don't receive an error, but instead a "200 OK" with body "OK", which is the same thing I would receive if resending the same data again by bypassing nginx. Kinda confusing, so I'm leaving the following to explain it better:

* data-1 -> nginx -> NodeBB = 400 Bad Request
* data-1 -> NodeBB = 202 Accepted
* data-1 -> NodeBB = 200 OK
* data-1 -> nginx -> NodeBB = 200 OK # as many times as you want, this will be the result
* data-2 -> nginx -> NodeBB = 400 Bad Request # if you were to change the data, the problem persists
* data-1 -> nginx -> NodeBB = 200 OK # but not for what was previously correctly processed

**Comment by julianlam** [comment]
To your point about the `200 OK`, this is because you've run afoul of the "history" middleware check.

When you made the call to NodeBB directly, it processed successfully, and returned `202`. When you made the same call via nginx, the call was routed appropriately to NodeBB but since it had already seen it (same `id`), it just send back a `200 OK`:

```
	// History check
	const seen = await db.isSortedSetMember('activities:datetime', req.body.id);
	if (seen) {
		// winston.verbose(`[middleware/activitypub] Activity already seen, ignoring (${req.body.id}).`);
		return res.sendStatus(200);
	}
```

**Comment by julianlam** [comment]
@barisusakli suggests that perhaps it is related to `proxy_redirect`. In our documentation, it is set to `proxy_redirect off;`. I noted that you've commented that out in your minimal nginx config

**Comment by andrigamerita** [comment]
> In our documentation, it is set to `proxy_redirect off;`. I noted that you don't actually have that line in your minimal nginx config

I've tried that, but it doesn't make a difference. As I've said in today's first messages, I've actually tried using the exact same configuration from the documentation, and yet that hasn't solved anything. I've actually tried tweaking the configuration right now with various options related to proxy and request body handling, but, again, no luck.

**Comment by julianlam** [comment]
Anyhow, in my limited testing it seems to be that the 400 is sent by NodeBB. That is, I've receiving `X-Powered-By: NodeBB`, which means the request is reaching NodeBB, but just is not being processed correctly.

Are you able to add debug logging to the running instance? That would be the quickest way to see what is going on.

**Comment by julianlam** [comment]
A cursory review of the codebase suggests that `400 Bad Request` is sent only when either:

1. the required properties are not present (this is not the case, as the history check comes _after_ this check)
2. HTTP signatures failed. In your logging do you see the `signature` header sent to NodeBB when the request is proxied through nginx? Keep in mind the signature is typically calculated through a couple other headers, so if _those_ values were changed in-transit, then the signature check will fail.

**Comment by julianlam** [comment]
Signature in your example above contains `headers="(request-target) host date digest content-type"`, so you need to look at those headers and compare. `(request-target)` is a special one that is something like `get https://example.org/new/inbox`

Hey, ,... maybe `(request-target)` changed because there's a subfolder? Though you mentioned this also occurs with non-subfolders.

**Comment by julianlam** [comment]
What I would suggest is adding a `config.log` to line ~50 of `src/middleware/activitypub.js`:

`const verified = await activitypub.verify(req);`

* Check the value of `verified`

If true, then add more logs to `src/activitypub/index.js`, around line 235, I'd check the value of `signed_string`, which is the reconstructed string that is later signed by the public key and compared.

**Comment by andrigamerita** [comment]
> Are you able to add debug logging to the running instance?

Forgot to mention that I had already tried that (with `./nodebb --log-level debug --dev start`) but, assuming I haven't misunderstood the command, it unfortunately never outputs any line regarding ActivityPub inbox errors (but I think the logging itself is working because it gives extra information on other matters, compared to reading the log without those extra flags, and it says verbose)

> Signature in your example above contains headers="(request-target) host date digest content-type", so you need to look at those headers and compare.

Well... I can't believe it. After the manual tests, and now you making me notice this, I've tried writing a line `proxy_set_header Host bbs.spacc.eu.org;` in the nginx config, to force the Host header having the forum's domain instead of being `127.0.0.1:4567` or `localhost:4567`.

So, the `Host` header being wrong was the problem all along and it would have sufficed to set this line in the configuration. Honestly don't know why `Host $http_host;` like indicated in the documentation hasn't worked. Probably because I actually specify 2 hosts in my real configuration, my clearweb domain and an additional Tor hidden service domain.
